### ISSUE TITLE
Clouds Server Crash

### ISSUE THREAD

**Issue by WenXin20** [ShetiPhian/Localizations-and-Issues#327]
**Describe the bug**
Sometimes my server crashes during the placement of clouds. FerriteCore is also mentioned in the log.

**Additional context**
Minecraft Version: 1.18.1
Forge Version: 39.0.79
ShetiPhian Core Version: 3.10.13
Terraqueous Version: 1.9.4
[FerriteCore](https://www.curseforge.com/minecraft/mc-mods/ferritecore) Version: 4.1.2
Are other mods present?
If yes:
- Does everything work without my mods? Yes
- Have you tested with only my mods? No

latest.log = https://mcpaste.io/2511997c55f51024
debug.log = n/a

**Comment by ShetiPhian** [comment]
I believe this is Ferrite.

- The crash is a threading issue, but I'm not toughing threads. Forge calls my generators to populate the chunk, and the cloud generator is mostly a copy of the Minecraft ore generator just replacing air rather then stone
- SmallThreadingDetector is part of FerriteCore
- The recent code change in FerriteCore: "Feature: Re-implement SmallThreadingDetector, this time in a way that doesn't crash when generating a world"

Now this still leaves the possibility I'm doing something unexpected that they didn't account for or maybe even can't

**Comment by WenXin20** [comment]
Ah ok, I'll report it there too.

**Comment by malte0811** [comment]
This has helped me find some small issues in my `ThreadingDetector` implementation, but all in the "crashing path", so nothing that should be able to cause a crash when there wouldn't have been a crash before. No idea if there are more issues lurking in it unfortunately. @WenXin20 can you try installing the FC build from [here](https://github.com/malte0811/FerriteCore/actions/runs/1882594080) (download the ZIP and extract the file from `forge/build/libs`)? And if it crashes again, upload a new crashlog?

It's annoying that the crash is so rare, but I've managed to [reproduce](https://gist.github.com/malte0811/056552cb1245c1b3f389bceba4d842c2#file-terraqueous-txt-L296-L311) a similar issue with FCs threading detector disabled (but the rest of it still loaded). What's interesting is that the OPs crash is a "proper" crash, while this (and what I've been getting) is a "worldgen is stuck" thing, which is hard to debug without `jstack` or similar.

I noticed one difference between your code and the vanilla `OreFeature`, but I have no idea whether it's related: Vanilla has an `ensureCanWrite` call before the `getSection`, your version doesn't.

**Comment by WenXin20** [comment]
Sure I'll give it a try, thanks.

**Comment by ShetiPhian** [comment]
Well this report has resulted in some interesting findings.

When I add ensureCanWrite, I start getting a "Detected setBlock in a far chunk" error.
Logically this tells me my code must be wrong.
But after several tests I just couldn't find the cause, so instead I recoded the clouds to use the ore generator, only providing it a configuration (what block to replace, what block state to place, how many blocks)
And with none of the generation itself controlled by me, I still got the "Detected setBlock in a far chunk" error. (and my clouds now looked bad)
Reverting back to my original code and implementing the logic of ensureCanWrite (without the log spam), appears to work until you look closely and notice clouds are cut off at chunk boundaries.

Something allows the ore generator to span chunks when it normally runs, but not while I'm reusing it for clouds.
My current guess is the decoration category as ores have their own, while I'm currently using the one after the surface is generated.

More tests needed...

**Comment by WenXin20** [comment]
> This has helped me find some small issues in my `ThreadingDetector` implementation, but all in the "crashing path", so nothing that should be able to cause a crash when there wouldn't have been a crash before. No idea if there are more issues lurking in it unfortunately. @WenXin20 can you try installing the FC build from [here](https://github.com/malte0811/FerriteCore/actions/runs/1882594080) (download the ZIP and extract the file from `forge/build/libs`)? And if it crashes again, upload a new crashlog?
> 
> It's annoying that the crash is so rare, but I've managed to [reproduce](https://gist.github.com/malte0811/056552cb1245c1b3f389bceba4d842c2#file-terraqueous-txt-L296-L311) a similar issue with FCs threading detector disabled (but the rest of it still loaded). What's interesting is that the OPs crash is a "proper" crash, while this (and what I've been getting) is a "worldgen is stuck" thing, which is hard to debug without `jstack` or similar.
> 
> I noticed one difference between your code and the vanilla `OreFeature`, but I have no idea whether it's related: Vanilla has an `ensureCanWrite` call before the `getSection`, your version doesn't.

I was able to get a new crash log
[latest.log](https://gist.github.com/WenXin20/e0891b9c5265db98860261b3a564acb5#file-latest-log-L4471)

**Comment by ShetiPhian** [comment]
So here is what I've got so far
https://www.mediafire.com/file/ksnqltuk1m38plf/terraqueous-1.18-1.9.5--DEBUG.jar/file

Changing the the decoration category didn't seem to do anything, but I left it as the ores due to using the ore generator.
I've reworked code and the generation is passed to the vanilla ore generator, this time via a different method where I can adjust additional parameters.
Either the clouds look the same, very similar or I've just forgot what they use to look like at this point lol

I could not completely get rid of the "Detected setBlock in a far chunk" error,  though its greatly reduced. one chunk in about every 120 just refuses to generate, but the ones around it will.
Since it isn't that often and things still look alright I've suppressed the log spam.
